src/main/common/log.c

```diff
@@ -0,0 +1,196 @@
+ *
+ * INAV is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with INAV.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <ctype.h>
+
+#include "build/version.h"
+
+#include "drivers/serial.h"
+#include "drivers/time.h"
+
+#include "common/log.h"
+#include "common/printf.h"
+#include "common/utils.h"
+
+#include "config/feature.h"
+#include "config/parameter_group_ids.h"
+
+#include "io/serial.h"
+
+#include "fc/config.h"
+
+#include "msp/msp.h"
+#include "msp/msp_serial.h"
+#include "msp/msp_protocol.h"
+
+#if defined(USE_LOG)
+
+#define LOG_PREFIX                  "[%6d.%03d] "
+#define LOG_PREFIX_FORMATTED_SIZE   13
+
+static serialPort_t * logPort = NULL;
+static mspPort_t * mspLogPort = NULL;
+
+PG_REGISTER(logConfig_t, logConfig, PG_LOG_CONFIG, 0);
+
+void logInit(void)
+{
+    const serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_LOG);
+    if (!portConfig) {
+        return;
+    }
+
+    bool portIsSharedWithMSP = false;
+
+    if (determinePortSharing(portConfig, FUNCTION_LOG) == PORTSHARING_SHARED) {
+        // We support sharing a LOG port only with MSP
+        if (portConfig->functionMask != (FUNCTION_LOG | FUNCTION_MSP)) {
+            return;
+        }
+        portIsSharedWithMSP = true;
+    }
+
+    // If the port is shared with MSP, reuse the port
+    if (portIsSharedWithMSP) {
+        const serialPort_t *logAndMspPort = findSharedSerialPort(FUNCTION_LOG, FUNCTION_MSP);
+        if (!logAndMspPort) {
+            return;
+        }
+
+        mspLogPort = mspSerialPortFind(logAndMspPort);
+        if (!mspLogPort) {
+            return;
+        }
+
+    } else {
+        logPort = openSerialPort(portConfig->identifier, FUNCTION_LOG, NULL, NULL, baudRates[BAUD_921600], MODE_TX, SERIAL_NOT_INVERTED);
+        if (!logPort) {
+            return;
+        }
+    }
+    // Initialization done
+    LOG_I_SYNC(SYSTEM, "%s/%s %s %s / %s (%s)",
+        FC_FIRMWARE_NAME,
+        targetName,
+        FC_VERSION_STRING,
+        buildDate,
+        buildTime,
+        shortGitRevision
+    );
+}
+
+static void logPutcp(void *p, char ch)
+{
+    *(*((char **) p))++ = ch;
+}
+
+static void logPrint(bool synchronous, const char *buf, size_t size)
+{
+    if (logPort) {
+        // Send data via UART (if configured & connected - a safeguard against zombie VCP)
+        if (serialIsConnected(logPort)) {
+            serialPrint(logPort, buf);
+            if (synchronous) {
+                waitForSerialPortToFinishTransmitting(logPort);
+            }
+        }
+    } else if (mspLogPort) {
+        mspSerialPushPort(MSP_DEBUGMSG, (uint8_t*)buf, size, mspLogPort, MSP_V2_NATIVE);
```

Comment:
This is actually dangerous. MSP on unconnected VCP will hang, could be very dangerous in flight.

Reply:
Ooops. Good catch! Thing is, we might also want sync messages when debugging to make the FC wait for the listener to connect. Maybe we can get rid of the _SYNC macros and add a setting to make all logging statements sync? In any case, I'll make the code path for synchronous logging check the arming state before blocking.